integration_tests/src/main/python/hash_aggregate_test.py

```diff
@@ -605,40 +605,47 @@ def test_subquery_in_agg(adaptive, expr):
         conf = {"spark.sql.adaptive.enabled" : adaptive})
 
 
+def assert_single_level_struct(df):
```

Comment:
The bug in Spark is about a value collisions. Spark will try to save space sending data to python and looks for `Row`s with duplicate values in them.  A `Row` is used to represent both a regular row as well as a Struct. So if I have a batch like

```
Row('a' = null)
Row('a' = Row('aa' = null))
```

There is still the possibility that this bug will be hit, even with no struct nesting, simply because a top level row in spark is treated like a struct so inherently all structs are nested under a top level row struct.  The best way to remove the possibility of hitting this bug is to have a different number of entries in each struct and also different from the rows.

Reply:
thanks you for explaining the Spark bug @revans2